c_src/exml_event.c

```diff
@@ -222,30 +273,57 @@ static ERL_NIF_TERM parse(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
     enif_get_int(env, argv[2], &is_final);
     enif_inspect_binary(env, argv[1], &stream);
 
-    expat_parser *parser_data = XML_GetUserData((XML_Parser)(*parser));
+    expat_parser *parser_data = XML_GetUserData(*parser);
     parser_data->result = enif_make_list(env, 0);
     parser_data->env = env;
-    XML_SetUserData((XML_Parser)(*parser), parser_data);
+    XML_SetUserData(*parser, parser_data);
 
-    res = XML_Parse((XML_Parser)(*parser), (const char *)stream.data, stream.size, is_final);
-    if (!res)
+    while (1)
         {
-            errcode = XML_GetErrorCode((XML_Parser)(*parser));
-            errstring = (char *)XML_ErrorString(errcode);
+            long start_index = get_current_byte_index(*parser);
+            res = XML_Parse(*parser, ((const char *)stream.data) + offset, stream.size - offset,
+                            is_final);
+            errcode = XML_GetErrorCode(*parser);
+
+            if (parser_data->start_tag == NULL)
+                break;
+
+            if (errcode == XML_ERROR_MISPLACED_XML_PI)
+                parser_data->restart_from = get_current_byte_index(*parser);
+            else if (errcode != XML_ERROR_ABORTED)
+                break;
+
+            parser_data->result =
+              enif_make_list_cell(parser_data->env, XML_STREAM_RESET, parser_data->result);
+            offset += parser_data->restart_from - start_index;
+            rest_parser_internal(env, *parser);
+        }
 
-            return enif_make_tuple(env, 2, ERROR, enif_make_string(env, errstring, ERL_NIF_LATIN1));
+    if (res == XML_STATUS_ERROR)
+        {
+            const char *errstring = (const char *)XML_ErrorString(errcode);
+            ERL_NIF_TERM reason = enif_make_string(env, errstring, ERL_NIF_LATIN1);
+            return enif_make_tuple(env, 2, ERROR, reason);
         }
 
     return enif_make_tuple(env, 2, OK, parser_data->result);
 };
 
 static int load(ErlNifEnv *env, void **priv, ERL_NIF_TERM info)
 {
+    if (sizeof(XML_Char) != sizeof(char))
+        {
+            fprintf(stderr, "XML_Char is not a typedef of char; the exml library will not work "
```

Comment:
I'm not sure what the reason is, but I noticed that to correctly align line breaks both `\n` and `\r` are needed in C format strings printed from .so files loaded by BEAM. Otherwise, we might get something like:

```
the printed string
                  ok
```

where `ok` is the return value printed by the shell. I'm not sure if `stderr`/`stdout` plays a role here.